flocker/node/agents/pd.py

```diff
@@ -0,0 +1,371 @@
+
+    :param unicode blockdevice_id: The blockdevice_id to get the dataset_id
+        for.
+
+    :returns UUID: The corresponding dataset_id.
+    """
+    return UUID(blockdevice_id[len(_PREFIX):])
+
+
+def _dataset_id_to_blockdevice_id(dataset_id):
+    """
+    Computes a blockdevice_id from a dataset_id.
+
+    :param UUID dataset_id: The dataset_id to get the blockdevice_id for.
+
+    :returns unicode: The corresponding blockdevice_id.
+    """
+    return _PREFIX + unicode(dataset_id)
+
+
+def _extract_attached_to(disk):
+    """
+    Given a GCE disk resource, determines the unicode name of the machine that
+    it is attached to.
+
+    :param dict disk: A GCE disk resource as returned from the API.
+
+    :returns: The `unicode` name of the instance the disk is attached to or
+        `None` if it is not attached to anything.
+    """
+    # TODO(mewert): determine how this works with a disk being attached to
+    # multiple machines, update comment above.
+    users = disk.get('users', [])
+    if not users:
+        return None
+    return unicode(users[0].split('/')[-1])
+
+
+@implementer(IBlockDeviceAPI)
+class PDBlockDeviceAPI(object):
+    """
+    A PD implementation of ``IBlockDeviceAPI`` which creates block devices in a
+    GCE project.
+
+    Constraints imposed from GCE:
+        - GCE does not have a grab-bag of metadata you can attach to disks.
+            Instead, it has two fields in the disk resource you can set:
+                - name (must be unique within a project,
+                        limited to 63 characters,
+                        must start with letter)
+                - description (free form text field)
+        - GCE does allow you to pick the blockdevice_id of your volumes (the
+            unique identifier that will be used to identify your volume on
+            subsequent API calls). In GCE terms this is the disk resource name.
+        - GCE lets you filter on both of these string fields, but only for
+            equals and not equals.
+        - GCE lets you set a token in the SCSI info pages on your blockdevices
+            every time you attach them to a VM so that symlinks are created at
+            '/dev/disk/by-id/google-<token>'.
+            - This technically is brittle to implementation, configuration, and
+                version of udev and the SCSI driver within the guest OS.
+        - GCE resource names and descriptions are immutable after creation.
+
+    Design:
+        - For a given dataset_id (06e07bcc-fea1-4810-8c6d-7487196998b6) prefix
+            the dataset_id with "flocker-v1-" to get the unicode blockdevice-id
+            (flocker-v1-06e07bcc-fea1-4810-8c6d-7487196998b6).
+        - Set the name of the GCE disk resource to the blockdevice_id. This
+            lets us perform operations on the disk only knowing its
+            blockdevice_id.
+        - Set the description of the disk resource to:
+            "flocker-v1-cluster: <cluster-uuid>".
```

Comment:
This isn't implemented yet.

I was thinking of sending this up for code review once this is implemented, and then have new PRs for the TODOs and adding more tests after that. Let me know if you think that is a bad plan.